storagemarket/impl/shardregistration.go

```diff
@@ -0,0 +1,146 @@
+package storageimpl
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/dagstore"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/extern/sector-storage/storiface"
+	"github.com/filecoin-project/specs-storage/storage"
+
+	mktdagstore "github.com/filecoin-project/go-fil-markets/dagstore"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-fil-markets/storagemarket/impl/providerstates"
+)
+
+var shardRegKey = datastore.NewKey("shards-registered")
+
+type SectorState interface {
+	StateSectorGetInfo(context.Context, address.Address, abi.SectorNumber, types.TipSetKey) (*miner.SectorOnChainInfo, error)
+	IsUnsealed(ctx context.Context, sector storage.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize) (bool, error)
+}
+
+type ShardRegistration struct {
+	maddr       address.Address
+	ds          datastore.Datastore
+	dagStore    mktdagstore.DagStoreWrapper
+	sectorState SectorState
+}
+
+func NewShardRegistration(
+	maddr address.Address,
+	ds datastore.Datastore,
+	dagStore mktdagstore.DagStoreWrapper,
+	sectorState SectorState,
+) *ShardRegistration {
+	return &ShardRegistration{
+		maddr:       maddr,
+		ds:          ds,
+		dagStore:    dagStore,
+		sectorState: sectorState,
+	}
+}
+
+func (r *ShardRegistration) registerShards(ctx context.Context, deals []storagemarket.MinerDeal) error {
+	// Check if all deals have already been registered as shards
+	has, err := r.ds.Has(shardRegKey)
+	if err != nil {
+		return xerrors.Errorf("failed to get shard registration status: %w", err)
+	}
+	if has {
+		// All deals have been registered as shards, bail out
+		return nil
+	}
+
+	// Filter for deals that are currently sealing.
+	// If the deal has not yet been handed off to the sealing subsystem, we
+	// don't need to call RegisterShard in this migration; RegisterShard will
+	// be called in the new code once the deal reaches the state where it's
+	// handed off to the sealing subsystem.
+	resch := make(chan dagstore.ShardResult, len(deals))
+	for _, deal := range deals {
+		if deal.Ref.PieceCid == nil {
+			continue
+		}
+
+		// Check if the deal has been handed off to the sealing subsystem
+		var sealing bool
+		for _, state := range providerstates.ProviderSealingStates {
+			if deal.State == state {
+				sealing = true
+				break
+			}
+		}
+		if !sealing {
+			continue
+		}
+
+		// Check if the deal is in an unsealed state
+		isUnsealed, err := r.isUnsealed(ctx, deal.SectorNumber)
+		if err != nil {
+			log.Errorf("failed to get unsealed state of deal with piece CID %s: %s", deal.Ref.PieceCid, err)
```

Comment:
Since we're not in control of the value of `isUnsealed` on error, let's be defensive here and set this to `false`, because we want shards for which we couldn't determine their unsealed status to be registered as lazy.